### PR TITLE
Normalize JSD distillation loss by num_items_in_batch for gradient accumulation

### LINKED ISSUE #4719: Loss calculation of `GKDTrainer` may be inaccurate when performing gradient accumulation?
It seems that `GKDTrainer` averages the loss of tokens in a micro batch ahead?

https://github.com/huggingface/trl/blob/8918c9836a3e0b43a6851c08d01b69072f56ca52/trl/experimental/gkd/gkd_trainer.py#L284

### PR DESCRIPTION
**Pull request by behroozazarkhalili**
## What

`GKDTrainer`, `GOLDTrainer`, and `DistillationTrainer` declared `num_items_in_batch` in `compute_loss` but never used it, so the JSD distillation loss was normalized by the **local per-microbatch** valid-token count instead of the **global** count.

## Why

Under gradient accumulation, transformers' `Trainer` skips its own `loss / gradient_accumulation_steps` correction precisely because the `compute_loss` signature accepts `num_items_in_batch` — but these trainers never applied the global normalization themselves. The result: accumulated microbatches were mis-weighted, and the loss (and gradients) depended on how a batch was split into microbatches. This is the same gradient-accumulation-correctness contract `SFTTrainer` already honors (`trl/trainer/sft_trainer.py`).

## What changed

Each trainer now honors `num_items_in_batch`, mirroring `SFTTrainer`:
- **When provided** (under gradient accumulation) → reduce as `sum / num_items_in_batch`.
- **When `None`** → keep the existing local `batchmean` (no behavior change on the non-accumulation path).

Applied across **every** JSD divergence path:
- **GKD / GOLD**: threaded through `generalized_jsd_loss` (the reduction now branches on `num_items_in_batch`), plus the Liger fused path (rescaled by `num_valid_local / num_items_in_batch`).
- **Distillation**: threaded through the shared `_reduce_divergence_loss` helper, so the full-vocabulary JSD, the **default** mixed top-1 path (`loss_top_k=1`), and the server paths are all covered — plus the Liger fused path.
- **GOLD's ULD path** is a separate loss with its own normalization and is intentionally left untouched.

## Tests

New `test_loss_normalizes_by_num_items_in_batch` in each trainer's test file asserts the grad-accum-correct relationship (`sum/N` reduction; doubling the global count halves the loss; with `N` equal to the local valid-token count the result matches the local mean). The Distillation test is parametrized over `loss_top_k ∈ {0, 1}` to cover both the full-vocabulary and default mixed-top-1 paths. A GPU-gated Liger test covers the fused path.

Resolves #4719

---

> [!NOTE]
> **Medium Risk**
> Changes loss scaling and which tokens contribute to JSD in core distillation trainers, which directly affects gradients under gradient accumulation and mixed prompt lengths; scope is limited to experimental trainers with regression tests.
> 
> **Overview**
> Fixes **gradient accumulation** for `GKDTrainer`, `GOLDTrainer`, and `DistillationTrainer` by actually using `num_items_in_batch` in JSD reduction (`sum / global count` when set; unchanged local `batchmean` when `None`), aligned with `SFTTrainer`. The Liger fused path is rescaled by `num_valid_local / num_items_in_batch`.
> 
> **GKD / GOLD** stop slicing logits and labels by batch-max `prompts` width; they use a standard causal shift and rely on `labels != -100` so variable-length prompts do not drop completion tokens. **GKD** `generate_on_policy_outputs` now masks prompt columns in `labels` with `-100`.
> 
> **Distillation** threads `num_items_in_batch` through `_reduce_divergence_loss` and local JSD / top-1 paths; **external teacher server** paths still normalize locally (documented follow-up). **GOLD ULD** is unchanged.
> 
> New tests cover `num_items_in_batch` scaling, variable-length prompt coverage, on-policy prompt masking, and Liger rescaling.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 45478ea7d8fde3874e48eabdf2e406786ae6c45f. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/gkd/gkd_trainer.py [hunk 1]
```diff
@@ -354,6 +366,14 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
                 teacher_bias=getattr(teacher_head, "bias", None),
             )
 
+            # The Liger JSD loss normalizes by the local number of valid tokens. Under gradient accumulation we want
+            # the global normalization, so rescale by `num_valid_local / num_items_in_batch`.
+            if num_items_in_batch is not None:
+                num_valid_local = (true_labels != -100).sum().clamp_min(1)
+                if isinstance(num_items_in_batch, torch.Tensor):
+                    num_items_in_batch = num_items_in_batch.to(loss.device)
+                loss = loss * num_valid_local / num_items_in_batch
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply trl/experimental/gkd/gkd_trainer.py:375]: Thanks! I looked into this — both paths actually yield exactly `0` for a microbatch with no valid labels, so there is no spurious contribution:

- **Non-Liger:** `jsd.sum()` over an empty mask is `0`, so `0 / num_items_in_batch == 0`.
- **Liger:** the fused kernel returns `soft_sum / num_valid_local`; with zero valid tokens `soft_sum ≈ 0`, and the rescale `loss * num_valid_local / num_items_in_batch` then multiplies a zero, so it stays `0`.

The `clamp_min(1)` is on `num_valid_local` — the local count used to undo the kernel's internal per-token normalization — not on the divisor `num_items_in_batch`. The divisor is left unclamped in **both** paths, which matches `SFTTrainer` (`loss = loss / num_items_in_batch`, no clamp). `num_items_in_batch` is the global valid-token count plumbed by `Trainer` and is positive by construction.

File: trl/experimental/distillation/distillation_trainer.py [hunk 2]
```diff
@@ -1460,12 +1484,14 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
...
                     completion_tokens=completion_tokens,
                     labels=trimmed_labels,
+                    num_items_in_batch=num_items_in_batch,
                 )
             else:
                 loss = self._compute_server_forward_kl_loss(
                     teacher_result=teacher_result,
                     student_log_probs=student_log_probs[:, :comp_len, :],
                     labels=trimmed_labels,
+                    num_items_in_batch=num_items_in_batch,
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply trl/experimental/distillation/distillation_trainer.py]: Good catch, thank you. You are right that on the teacher-server path the divergence is summed over the trimmed teacher window while `num_items_in_batch` counts every valid student label in the full batch, so the two token sets can differ and the global count would be the wrong denominator there.

Fixed in 400cc7b8 by keeping the server path on its local `batchmean` normalization (it no longer receives `num_items_in_batch`). This is self-consistent by construction: numerator and denominator are both over the same trimmed teacher window, regardless of how far the teacher logprobs reach. The server path therefore does not get gradient-accumulation normalization in this PR; I left that as a follow-up, since the global count of teacher-scored tokens cannot be recovered inside a single `compute_loss` call (HF Trainer computes `num_items_in_batch` upstream over untrimmed labels).

The local and Liger paths normalize over the same untrimmed labels that `num_items_in_batch` counts, so they keep the gradient-accumulation fix.

**Comment by behroozazarkhalili** [comment]
## Follow-up fix in `a418df0d`: the manual JSD path was skipping completion tokens when prompts have different lengths

**In one sentence:** the non-Liger GKD/GOLD loss only counted completion tokens that came *after the longest prompt in the batch*, so shorter-prompt samples had some of their completion tokens silently dropped from the loss.

### What went wrong

In a batch, each sample has a different prompt length. The collator pads two things to two **different** widths:

- `labels` is padded to the **full sequence** length (prompt + completion).
- `prompts` is padded to the **prompt-only** length.

The old code sliced the loss using the prompt width:

```python
prompt_lengths = inputs["prompts"].shape[1]   # = the LONGEST prompt in the batch
shifted_labels = inputs["labels"][:, prompt_lengths:]
```

For a sample whose prompt is shorter than the batch's longest prompt, its completion starts *before* column `prompt_lengths`. Slicing at `prompt_lengths` therefore cuts off the beginning of that sample's completion.

### Concrete example

A 2-sample batch (short prompt + long prompt):

```
labels.shape      = (2, 44)        # full-sequence width
prompts.shape[1]  = 41             # longest prompt -> the column the old code sliced at
sample 0 completion actually starts at column 35   # but is cut at 41 -> 6 tokens lost
```

| | tokens the loss summed | tokens `num_items_in_batch` counts |
|---|---|---|
| before fix | **9** | 15 |
| after fix | **15** | 15 |

Dividing a 9-token loss by 15 gives the wrong scale. Measured on the same batch:

| | local mean | sum / num_items | match? |
|---|---|---|---|
| before fix | 0.25855 | 0.12928 | ❌ (off by ~2x) |
| after fix | 0.26761 | 0.26761 | ✅ |

### The fix

Use the normal causal shift and let the existing `labels != -100` mask handle prompt exclusion (it already does):

```python
shifted_student_logits = student_outputs.logits[:, :-1, :]
shifted_teacher_logits = teacher_outputs.logits[:, :-1, :]
shifted_labels         = inputs["labels"][:, 1:]
```

No prompt-width slice is needed because prompt and padding positions are already `-100` in `labels`. This is exactly what the Liger path already does.

### Why this is safe for existing tests

When all prompts in a batch are the **same** length, the old slice and the new shift select the identical tokens, so the loss value is unchanged (verified: both give `jsd_sum = 0.7104` over the same 6 tokens). The bug only appeared with **variable-length** prompts.

The new test `test_loss_covers_all_completion_tokens_with_variable_length_prompts` builds a variable-length-prompt batch and checks the loss covers every completion token. It fails on the old code and passes with the fix.

File: trl/experimental/gkd/gkd_trainer.py [hunk 3]
```diff
@@ -376,18 +396,22 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
...
+            shifted_teacher_logits = teacher_outputs.logits[:, :-1, :]
+            shifted_labels = inputs["labels"][:, 1:]
 
             # compute loss
             loss = self.generalized_jsd_loss(
                 student_logits=shifted_student_logits,
                 teacher_logits=shifted_teacher_logits,
                 labels=shifted_labels,
                 beta=self.beta,
+                num_items_in_batch=num_items_in_batch,
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply trl/experimental/gkd/gkd_trainer.py:414]: Good catch, thank you. You are right: in the on-policy (`lmbda`) and seq-KD paths, `training_step` replaces the labels via `generate_on_policy_outputs`, which cloned the full generated sequence and masked only padding. The prompt tokens stayed as valid labels, so once `compute_loss` stopped slicing by prompt width and started relying on the `labels == -100` mask, the JSD loss was applied to prompt positions there.

Fixed in 46882555 by masking the prompt region in `generate_on_policy_outputs`:

```python
prompt_length = inputs["prompts"].shape[1]
new_labels[:, :prompt_length] = -100
```

`generate` echoes the prompt back as the first `prompt_length` columns, so this masks exactly the prompt and matches the collator convention (`labels[:len(prompt)] = -100`) that `compute_loss` relies on. It also brings GKD into line with the GOLD and Distillation on-policy paths, which already build their labels completion-only.

Verified with a regression test (`test_generate_on_policy_outputs_masks_prompt`): it uses variable-length prompts (so batched left-padding is exercised) and `min_new_tokens` to force a non-trivial completion, then asserts the prompt columns are all `-100` and the completion region still carries signal. The test fails on the previous code and passes with the fix.

**Comment by kashif** [comment]
good find @behroozazarkhalili checking too!

**Review by kashif — APPROVED** [review]
